Add read/write connection configurations

othercorey opened the pull request:
closes https://github.com/cakephp/cakephp/issues/16579

This uses a connection naming convention to switch between read and write connections. 

Read connections have a `:read` suffix. A `:read` connection  must have a matching write connection.

Calling `useReadRole()` or `useWriteRole()` on a query switches the connection instances at that point.

othercorey committed:
Add read and write connection config support

markstory reviewed (commented):
I like this. It is a simple solution that allows users to start using replicas. We're providing a convention for connections and a transparent easy to identify/find API for using replicas :clap:

I don't think additional magic for this is helpful. Using a replica requires contextual knowledge of an application and its consistency requirements which we simply don't have at the framework level.

othercorey commented:
I thought about adding a `$role` parameter to ConnectionManager::get(), but I didn't want to default it to write. We could default it to null, but not sure if that would be confusing.

markstory commented on src/Database/Connection.php:
```diff
@@ -177,6 +177,16 @@ public function configName(): string
     }
 
+    /**
+     * Returns the connetion role: read or write.
```
```suggestion
     * Returns the connection role: read or write.
```

markstory commented on tests/TestCase/Database/ConnectionTest.php:
```diff
@@ -1355,4 +1357,16 @@ public function testNoAutomaticReconnect(): void
+
+    public function testRoles(): void
+    {
+        $this->assertSame('write', $this->connection->role());
```
Should we use the constant here?

othercorey replied:
Yea ,we can. I was doing a sanity check that I set the names properly everywhere.

othercorey commented:
@markstory I added a `$role` parameter to `ConnectionManager::get()` to allow saving a role name. Made `getRead/WriteName()`  protected and added `getName()` wrapper.

Updated `Query::useRead/WriteRole()` to only switch if the role doesn't match expected. Added `useRole()` to switch based on role name.

Added test for exception thrown for missing write connection in setConfig().

markstory reviewed (commented):
Looks good to me.
